src/mcu/reflash.cpp

```diff
@@ -0,0 +1,441 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ * Copyright (C) 2022, KNS Group LLC (YADRO).
+ */
+#include "backplane_mcu_driver.hpp"
+#include "common/mmapfile.hpp"
+
+#include <gpiod.hpp>
+#include <nlohmann/json.hpp>
+
+#include <chrono>
+#include <cstdio>
+#include <filesystem>
+#include <fstream>
+#include <thread>
+
+namespace fs = std::filesystem;
+
+/**
+ * @brief Check MCU and reflash if required
+ *
+ * @param bus      - i2c-bus that MCU is located on
+ * @param addr     - MCU's i2c address
+ * @param force    - Interpret the MCU absence as an error
+ * @param firmware - Path to firmware image
+ * @param version  - Required frimware version
+ */
+static void updateMCU(int bus, int addr, bool force, const fs::path& firmware,
+                      const std::string& version)
+{
+    std::string dev = "/dev/i2c-" + std::to_string(bus);
+    std::unique_ptr<BackplaneMCUDriver> mcu;
+    try
+    {
+        mcu = backplaneMCU(dev, addr);
+    }
+    catch (const std::exception& e)
+    {
+        if (force)
+        {
+            fprintf(stderr, "MCU_%d_%02X: unable to init, %s\n", bus, addr,
+                    e.what());
+        }
+        return;
+    }
+
+    try
+    {
+        auto devType = mcu->getBoardType();
+        auto fwVer = mcu->getFwVersion();
+
+        printf("MCU_%d_%02X: type='%s', ver='%s'\n", bus, addr, devType.c_str(),
+               fwVer.c_str());
+
+        if (!version.empty() && version == fwVer)
+        {
+            printf("MCU_%d_%02X is running on the same version, reflashing "
+                   "skipped.\n",
+                   bus, addr);
+            return;
+        }
+    }
+    catch (const std::exception& e)
+    {
+        fprintf(stderr, "MCU_%d_%02X: Unable to query, %s\n", bus, addr,
+                e.what());
+        return;
+    }
+
+    try
+    {
+        if (!firmware.empty())
+        {
+            auto fw = common::MappedMem::open(firmware);
+
+            static constexpr size_t minImageSize =
+                64; // image can't be less then 64 bytes (header size)
+            static constexpr size_t maxImageSize =
+                128 * 1024; // image can't be more then 128 Kbytes
+            if ((fw.size() < minImageSize) || (fw.size() > maxImageSize))
+            {
+                fprintf(stderr, "Incorrect '%s' size\n", firmware.c_str());
+                return;
+            }
+
+            mcu->eraseFlash();
```

Comment:
I thought you wanted to skip this

Reply:
That happened later that this pull-request created. :)
Fixed.